### PR TITLE
Edit workflow

### PR DESCRIPTION
**Pull request by GGn0**
The workflow is split into three sections (jobs):
- build
- stage
- deploy

### Build step

Is always executed  
It tries to export the static site and caches the result

### Stage step

Runs after build only on pull requests
The build results are pushed to heroku (see the configuration below)
When the step finishes, a comment is added on the PR page containing links to the heroku app and the deploy approval page  

image

### Deploy step

Runs after stage only on pull requests
The workflow is paused here.
The user needs to review the staged site on heroku and approve it to run this step  

image

Once the deploy step finishes, the site is still pushed to the branch gh-pages

### Configuration
Since the workflow uses heroku to deploy the site during the staging step you will need to:
- Create an account on heroku (If you don't have one already 😄)
- Create an app on heroku
- Setup repository secrets
image
  - HEROKU_EMAIL: Your heroku account email
  - HEROKU_API_KEY: Your heroku api key that can be found at [dashboard.heroku.com/account](https://dashboard.heroku.com/account)
  - HEROKU_STAGING_APP: The name of the app created before

### CODE REVIEW

**Comment by GGn0** [comment]
@gmacario Could you try adding the repository secrets and re-running the jobs?

**Comment by gmacario** [comment]
@GGn0 one check failed  - PTAL
https://github.com/gmacario/gmacario.github.io/pull/109/checks?check_run_id=2709476670

File: .github/tests/package.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "serve-static-site",
+  "version": "1.0.0",
+  "description": "Serve a static site withanhttp server",
```
Comment (gmacario) [inline .github/tests/package.json:4]: s/withanhttp/with a http/

File: .github/workflows/build-and-deploy.yml [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    
+    defaults:
+      run:
+        shell: bash
+        working-directory: ./src
+    
+    steps:
+    
+      - name: Checkout 🛎️
+        uses: actions/checkout@v2 # If you're using actions/checkout@v2 you must set persist-credentials to false in most cases for the deployment to work correctly.
```
Comment (gmacario) [inline .github/workflows/build-and-deploy.yml:70]: Split lines > 120 chars

File: .github/workflows/build-and-deploy.yml [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      run:
+        shell: bash
+        working-directory: ./src
+        
+    environment:
+      name: staging_environment  
+      
+    steps:          
+      - name: Checkout 🛎️
+        uses: actions/checkout@v2 # If you're using actions/checkout@v2 you must set persist-credentials to false in most cases for the deployment to work correctly.
```
Comment (gmacario) [inline .github/workflows/build-and-deploy.yml:147]: Split lines > 120 chars

File: .github/workflows/build-and-deploy.yml [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        working-directory: ${{ github.workspace }}
+       
+      - name: Cache build artifacts 📦
+        if: ${{env.deploy_staging == 'true'}}
+        uses: actions/cache@v2
+        with:
+          path: ${{ github.workspace }}/_static
+          key: ${{ runner.os }}-static-site-${{ github.sha }}
+  
+  stage:
```
Comment (gmacario) [inline .github/workflows/build-and-deploy.yml:53]: s/stage/deploy_staging/

File: .github/workflows/build-and-deploy.yml [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        working-directory: ./src
+    
+    steps:
+    
+      - name: Checkout 🛎️
+        uses: actions/checkout@v2 # If you're using actions/checkout@v2 you must set persist-credentials to false in most cases for the deployment to work correctly.
+        with:
+          persist-credentials: false
+          
+      - name: Setup heroku CLI 🛠
```
Comment (gmacario) [inline .github/workflows/build-and-deploy.yml:74]: s/Setup heroku CLI/Setup Heroku CLI/

Reference: [Heroku is capitalized or not?](https://www.google.com/search?q=heroku+is+capitalized+or+not&rlz=1C1SQJL_itIT952IT952&sxsrf=ALeKk00-PAUNYHvZpI_fxumo8zz4V2DpVw%3A1622613777652&ei=ER-3YOemJ86I9u8P-Zi9iAg&oq=heroku+is+capitalized+or+not&gs_lcp=Cgdnd3Mtd2l6EAM6BwgAEEcQsAM6BQghEKABOggIIRAWEB0QHlC-KljjMGCTNmgBcAJ4AIABc4gBvAWSAQM3LjGYAQCgAQGqAQdnd3Mtd2l6yAEIwAEB&sclient=gws-wiz&ved=0ahUKEwin88uuo_jwAhVOhP0HHXlMD4EQ4dUDCA8&uact=5)

File: .github/workflows/build-and-deploy.yml [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+          mkdir heroku
+        working-directory: ${{ github.workspace }}
+          
+      - name: Restore build artifacts cache 📦
+        uses: actions/cache@v2
+        with:
+          path: ${{ github.workspace }}/_static
+          key: ${{ runner.os }}-static-site-${{ github.sha }}
+      
+      - name: Deploy to heroku 🚀
```
Comment (gmacario) [inline .github/workflows/build-and-deploy.yml:94]: s/heroku/Heroku/

See above

File: .github/workflows/build-and-deploy.yml [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+          git config --global user.email "${heroku_email}"
+          git config --global user.name "Github Action"
+          heroku git:remote -a ${heroku_app}
+          heroku login
+          git add -A
+          git commit -am "make it better"
+          git push heroku master --force
+        working-directory: ${{ github.workspace }}
+        
+      - name: Create comment on pr 🧾
```
Comment (gmacario) [inline .github/workflows/build-and-deploy.yml:115]: s/Create comment on pr/Create comment on PR/

File: .github/workflows/build-and-deploy.yml [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+          cd heroku
+          cat >.gitignore <<EOF
+          node_modules/
+          EOF
+          cat >Procfile <<EOF
+          web: npm install && npm start
+          EOF
+          git init
+          git config --global user.email "${heroku_email}"
+          git config --global user.name "Github Action"
```
Comment (gmacario) [inline .github/workflows/build-and-deploy.yml:107]: s/Github Action/GitHub Action/

**Comment by GGn0** [comment]
> @GGn0 one check failed - PTAL
> https://github.com/gmacario/gmacario.github.io/pull/109/checks?check_run_id=2709476670
> 
> Update: just read your previous comment... let me take a look, then
> 
> Anyway IMHO the error message around the failure is a little bit misleading:
> 
> ```
> Run cp -r .github/tests/* heroku
> hint: Using 'master' as the name for the initial branch. This default branch name
> hint: is subject to change. To configure the initial branch name to use in all
> hint: of your new repositories, which will suppress this warning, call:
> hint: 
> hint: 	git config --global init.defaultBranch <name>
> hint: 
> hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
> hint: 'development'. The just-created branch can be renamed via this command:
> hint: 
> hint: 	git branch -m <name>
> Initialized empty Git repository in /home/runner/work/gmacario.github.io/gmacario.github.io/heroku/.git/
>  ›   Error: Flag --app expects a value
> Error: Process completed with exit code 2.
> ```
> 
> I would try to make some checks and print some more meaningful errors instead

Absolutely
I'll look into adding some more checks

**Comment by GGn0** [comment]
I added a step in the workflow to check if the right secrets have been set
image

**Comment by gmacario** [comment]
Now that #113 has been merged I believe this PR may be closed.

**Pull request closed without merging**